modules/extensions/src/terrain/terrain-effect.ts

```diff
@@ -0,0 +1,238 @@
+import {Texture2D, ProgramManager, Framebuffer} from '@luma.gl/core';
+import {log} from '@deck.gl/core';
+
+import {terrainModule, TerrainModuleSettings} from './shader-module';
+import {TerrainCover} from './terrain-cover';
+import {TerrainPass} from './terrain-pass';
+import {TerrainPickingPass, TerrainPickingPassRenderOptions} from './terrain-picking-pass';
+import {createRenderTarget} from './utils';
+
+import type {Effect, PreRenderOptions, Layer, Viewport} from '@deck.gl/core';
+
+/** Class to manage terrain effect */
+export class TerrainEffect implements Effect {
+  id = 'terrain-effect';
+  props = null;
+  useInPicking = true;
+
+  /** true if inside a picking pass */
+  private isPicking: boolean = false;
```

Comment:
Given that `useInPicking` is set on `Effect`, would it not be better to move `isPicking` there also and then have it set externally rather than inferring from the Pass?

Reply:
Technically this flag is not a property of the effect, but a property of the current pass. What we need is an `isPicking` in `PreRenderOptions`, just like in `LayerFilterContext`. I'm not making that change here because it's going to be a much more extensive API change. All these bits of information about the current render pipeline are scattered and slightly different when passed to each effect method, layer method, pass method, and user callbacks. We need to do a thorough review of the Effect interface in v9.0, now that we have a lot more use cases to consider than when it was first designed.

Given that we are very close to the 8.9 release, I propose that we don't rush to another big API overhaul prematurely. At least this is a self-contained module and will be easy to refactor when we're ready.